// kpi-embed/vite.config.ts
// kpi/vite.config.ts
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'
import path from 'path'
import { fileURLToPath } from 'url'
const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

export default defineConfig(({ mode }) => {
  const isDevBuild = mode === 'development'
  return {
    plugins: [react()],
    define: {
      'process.env.NODE_ENV': JSON.stringify(isDevBuild ? 'development' : 'production'),
      global: 'window',
    },
    resolve: { alias: { '@': path.resolve(__dirname, 'src') } },
    build: {
      sourcemap: isDevBuild,
      minify: isDevBuild ? false : 'esbuild',
      lib: {
        entry: 'src/widget.tsx',
        name: 'KpiWidget',
        fileName: (format) => isDevBuild ? `kpi-widget.dev.${format}.js` : `kpi-widget.${format}.js`,
        formats: ['es', 'umd'],
      },
      rollupOptions: {
        external: ['react', 'react-dom', 'react-dom/client'],
        treeshake: !isDevBuild,
        output: {
          globals: {
            react: 'React',
            'react-dom': 'ReactDOM',           // 👈 portal
            'react-dom/client': 'ReactDOMClient', // 👈 createRoot
          },
          exports: 'named',
        },
      },
    },
  }
})
